File: app/Http/Controllers/SpecialistPatternsJava/Lab1/Lab1Java/src/graph/GraphObject.java
package graph;

// Абстрактный класс GraphObject представляет собой абстрактный графический объект.
public abstract class GraphObject implements Cloneable {
    public static final String DEFAULT_COLOR = "black"; // Константа для цвета по умолчанию.
    private String color; // Приватное поле для хранения цвета объекта.

    // Конструктор по умолчанию, инициализирующий объект с цветом по умолчанию.
    public GraphObject() {
        this(DEFAULT_COLOR);
    }

    // Конструктор, позволяющий инициализировать объект с заданным цветом.
    public GraphObject(String color) {
        super();
        this.color = color;
    }

    // Метод для получения цвета объекта.
    public String getColor() {
        return color;
    }

    // Метод для установки цвета объекта.
    public void setColor(String color) {
        this.color = color;
    }

    // Абстрактный метод для отрисовки объекта, который должен быть реализован в подклассах.
    public abstract void draw();
}

File: app/Http/Controllers/SpecialistPatterns/LR/LR1SP/graph/GraphObject.php
<?php

namespace App\Http\Controllers\SpecialistPatterns\LR\LR1SP\graph;

// Абстрактный класс GraphObject представляет собой абстрактный графический объект.
abstract class GraphObject
{
    // Константа для хранения цвета по умолчанию.
    public const DEFAULT_COLOR = "black";
    // Приватное свойство для хранения цвета объекта.
    private string $color;

    // Конструктор класса. Инициализирует объект заданным цветом, либо цветом по умолчанию.
    public function __construct(string $color = self::DEFAULT_COLOR)
    {
        $this->color = $color;
    }

    // Метод для получения цвета объекта.
    public function getColor(): string
    {
        return $this->color;
    }

    // Метод для установки цвета объекта.
    public function setColor(string $color): void
    {
        $this->color = $color;
    }

    // Абстрактный метод для отрисовки объекта, который должен быть реализован в подклассах.
    abstract public function draw(): void;

    // Метод для создания копии объекта.
    public function clone(): self
    {
        // Создаем новый объект того же класса и копируем значения свойств, включая цвет.
        $newObject = new static($this->color);
        // Можно добавить дополнительные операции копирования свойств, если необходимо.
        // $newObject->someProperty = $this->someProperty;
        return $newObject;
    }
}


